lib/index.js

```diff
@@ -309,4 +324,236 @@ module.exports = {
 
     return debugTree(new MergeTrees([ember, templateCompiler, jquery]), 'vendor:final');
   },
+
+  _issueGlobalsDeprecation() {
+    if (process.env.EMBER_ENV === 'production') {
+      return;
+    }
+
+    let isYarnProject = ((root) => {
+      try {
+        // eslint-disable-next-line node/no-unpublished-require
+        return require('ember-cli/lib/utilities/is-yarn-project')(root);
+      } catch {
+        return undefined;
+      }
+    })(this.project.root);
+
+    let groupedByTopLevelAddon = Object.create(null);
+    let groupedByVersion = Object.create(null);
+    let projectInfo;
+
+    for (let [addon, pathToAddon] of walkAddonTree(this.project)) {
+      let version = addon.pkg.version;
```

Comment:
We may  (speculation, and untested) want to skip the current iteration if the `pathToAddon` has already appeared. Some large applications, especially those using engines can have a very LARGE number of add-ons that must be considered (10s or 100s of thousands), and it is often the case that these extra traversals can cause CLI boot-time regressions.

But if the `pathToAddon` has already been considered it is safe to use the previous result, and skip the upcoming work.